lib/upnp/upnp.go

```diff
@@ -102,13 +116,19 @@ func Discover(ctx context.Context, _, timeout time.Duration) []nat.Device {
 			continue
 		}
 
-		for _, deviceType := range []string{"urn:schemas-upnp-org:device:InternetGatewayDevice:1", "urn:schemas-upnp-org:device:InternetGatewayDevice:2"} {
-			wg.Add(1)
-			go func(intf net.Interface, deviceType string) {
-				discover(ctx, &intf, deviceType, timeout, resultChan)
-				wg.Done()
-			}(intf, deviceType)
-		}
+		wg.Add(1)
+		// Discovery is done sequentially per interface because we discovered that FritzBox routers
+		// return a broken result sometimes if the IPv4 and IPv6 request arrive at the same time.
+		go func(iface net.Interface) {
+			// Discover IPv6 gateways on interface. Only discover IGDv2, since IGDv1 + IPv6 is not standardized and will lead to duplicates on routers.
+			discover(ctx, &iface, urnIgdV2, timeout, resultChan, true)
```

Comment:
Should we skip IPv6 discovery if the interface doesn't have a global unicast address?

Reply:
I'm not sure.  What exactly do you mean with "IPv6 discovery"?  Gateways reachable via IPv6?  I don't see why we should not use them, even if just to register an IPv4 mapping.  Doesn't hurt at least?

Reply:
If you discover an IPv6 gateway, it will advertize an IPv6 URL to interact with it. UPnP security (if present at all) is usually source based: IP address X is allowed to open ports for itself.

I wouldn't trust router vendors to properly support the case of "IPv6 X is allowed to open ports for IPv4 Y".

The idea behind my question is that we could skip a lot of pointless SSDP search requests. If there's no global unicast address on the interface, than it's not facing an IPv6 capable router.

Even an IPv4-only environment usually has a link-local IPv6 for each interface.

Reply:
This is implemented now